[opaque obj] Add value-type opaque obj

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169846
* __->__ #167768

Opaque objects are the way we allow custom operators to accept a user-defined "black box" object as an input. There are two kinds of opaque types: VALUE type and REFERENCE type. The distinction determines how torch.compile handles the object.

REFERENCE TYPES (default):

Reference-typed opaque objects represent mutable stateful objects and are
treated as black boxes. In torch.compile, since torch.compile cannot optimize
the anything (including tensors) within the object, the object must be an
input to the graph.
You can register a custom class as being a reference-based opaque object class
through `register_opaque_type(MyClass)`.

VALUE TYPES:

Value-typed opaque objects represent constant values.
In torch.compile, the graph specializes on the object like how other constants
are. Therefore there are a couple of methods on the class that must be
implemented before registering it as a value-typed opaque object class:
  - `__eq__`: torch.compile will create guards based on the equality of this
  object, meaning that a recompilation will happen if `__eq__` returns False.
  - `__hash__`: This must be implemented for Fake Tensor caching
  - `__repr__`: This must be implemented as it will be used in the FX graph's
    codegen to reconstruct the object. The string representation must be able to
    construct the object again through its `__init__` method.
You can register a custom class as being a reference-based opaque object class
through `register_opaque_type(MyClass, value_type=True)`.

The graph containing a value-type looks like:
```
def forward(self, L_x_ : torch.Tensor):
    l_x_ = L_x_
    process_with_config = torch.ops._TestOpaqueObject.process_with_config(l_x_, ValueConfig(mode='square'));  l_x_ = None
    return (process_with_config,)
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @chenyang78

angelayi committed:
[opaque obj] Add value-type opaque obj

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Add value-type opaque obj"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zou3519 commented on test/test_opaque_obj_v2.py:
```diff
@@ -526,6 +594,136 @@ def forward(self, primals, tangents):
+    cat = torch.ops.aten.cat.default([arg0_1, ones]);  arg0_1 = ones = None
+    add = torch.ops.aten.add.Tensor(cat, 3);  cat = None
+    return (add,)""",  # noqa: B950
+        )
```
does backend="inductor" work? I don't see a test for it

zou3519 commented on test/test_opaque_obj_v2.py:
```diff
@@ -70,10 +70,47 @@ def forward(self, x, y):
+
+# Register these types as a pytree constant to make it a value-type
+torch.utils._pytree.register_constant(ValueConfig)
+torch.utils._pytree.register_constant(SizeStore)
```
also I am confused because pytree.register_constant uses flat_apply but I don't see any flat_apply in the graphs below

zou3519 replied:
Can we kill pytree.register_constant?

zou3519 commented on torch/_library/opaque_object.py:
```diff
@@ -1,10 +1,36 @@
+# Note [Opaque Objects]
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# Opaque objects are objects that we allow to be passed into custom operators.
```
"We allow users to specify that custom operators may accept their own user-defined opaque objects as inputs"

zou3519 commented on torch/fx/graph.py:
```diff
@@ -562,6 +563,18 @@ def _get_repr(arg: Any) -> str:
+                        "expected to have to have a non-default `__repr__` "
+                        "implementation."
+                    )
+                arg_repr = repr(arg)
```
nvm I see the writeup, but the second point still stands

zou3519 commented on torch/_library/opaque_object.py:
```diff
@@ -1,10 +1,36 @@
+# - When torch.compiling, the object is treated as a constant and the graph
+#   specializes on the constant. Guards are based on equality (obj.__eq__),
+#   meaning torch.compile will recompile if the hash of the object changes.
+# - The type's __eq__, __hash__, and __repr__ must be implemented.
```
This note is basically a design note, so.. the question is why do these need to be implemented and what are they used for?

zou3519 reviewed (commented):
got questions, I think the main ones are:
1) is this the right way of rationalizing opaque value types with pytree.register_constant? 
2) let's make the note clearer, it's basically the design doc

angelayi committed:
Update on "[opaque obj] Add value-type opaque obj"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

XuehaiPan commented on torch/_library/opaque_object.py:
```diff
@@ -5,6 +5,42 @@
+#
+# You can register a custom class as being a reference-based opaque object class
+# through `register_opaque_type(MyClass, value_type=True)`.
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
nit: Maybe this can be a docstring at the top of the file instead of a block comment.


```python
"""Opaque Objects

...
"""

...
```

angelayi committed:
Update on "[opaque obj] Add value-type opaque obj"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela chenyang78

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Add value-type opaque obj"


Opaque objects are the way we allow custom operators to accept a user-defined "black box" object as an input. There are two kinds of opaque types: VALUE type and REFERENCE type. The distinction determines how torch.compile handles the object.

REFERENCE TYPES (default):

Reference-typed opaque objects represent mutable stateful objects and are
treated as black boxes. In torch.compile, since torch.compile cannot optimize
the anything (including tensors) within the object, the object must be an
input to the graph.
You can register a custom class as being a reference-based opaque object class
through `register_opaque_type(MyClass)`.

VALUE TYPES:

Value-typed opaque objects represent constant values.
In torch.compile, the graph specializes on the object like how other constants
are. Therefore there are a couple of methods on the class that must be
implemented before registering it as a value-typed opaque object class:
  - `__eq__`: torch.compile will create guards based on the equality of this
  object, meaning that a recompilation will happen if `__eq__` returns False.
  - `__hash__`: This must be implemented for Fake Tensor caching
  - `__repr__`: This must be implemented as it will be used in the FX graph's
    codegen to reconstruct the object. The string representation must be able to
    construct the object again through its `__init__` method.
You can register a custom class as being a reference-based opaque object class
through `register_opaque_type(MyClass, value_type=True)`.

The graph containing a value-type looks like:
```
def forward(self, L_x_ : torch.Tensor):
    l_x_ = L_x_
    process_with_config = torch.ops._TestOpaqueObject.process_with_config(l_x_, ValueConfig(mode='square'));  l_x_ = None
    return (process_with_config,)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela chenyang78

[ghstack-poisoned]

zou3519 commented on test/test_opaque_obj_v2.py:
```diff
@@ -70,10 +72,43 @@ def forward(self, x, y):
 register_opaque_type(RNGState)
 register_opaque_type(Counter)
 register_opaque_type(AddModule)
+register_opaque_type(ValueConfig, value_type=True)
```
nit: I prefer prefer typ="value"/"reference" over `value_type=True`, this implies that a non-value type is always a reference type (which I'm not sure is true in general?)

zou3519 commented on test/test_opaque_obj_v2.py:
```diff
@@ -553,6 +619,157 @@ def test_invalid_schema(self):
+
+        register_opaque_type(Bad2)
+        with self.assertRaisesRegex(ValueError, "cannot be registered as a pytree"):
+            pytree.register_dataclass(Bad2)
```
what's the rule here? something can only be registered as either a pytree or opaque object but not both?

If so, can include more of the error message or leave a comment

angelayi replied:
yes, we cant register it as a pytree if it's already an opaque object and vise versa. updated the error msg!

zou3519 commented on test/test_opaque_obj_v2.py:
```diff
@@ -553,6 +619,157 @@ def test_invalid_schema(self):
+        with self.assertRaisesRegex(
+            ValueError, "cannot be registered as an opaque object"
+        ):
+            register_opaque_type(Bad1)
```
do you handle the case where we do register_opaque_type(Foo) and then Foo goes out of scope correctly? The footgun to avoid is that we shouldn't use id(Foo) in a dict somewhere, and we also probably want a weakref to Foo (instead of a strong reference to it)

zou3519 reviewed (commented):
torch/_library/autograd.py changes look suspicious so I didn't approve this yet. Everything else LGTM with minor comments

angelayi committed:
Update on "[opaque obj] Add value-type opaque obj"


Opaque objects are the way we allow custom operators to accept a user-defined "black box" object as an input. There are two kinds of opaque types: VALUE type and REFERENCE type. The distinction determines how torch.compile handles the object.

REFERENCE TYPES (default):

Reference-typed opaque objects represent mutable stateful objects and are
treated as black boxes. In torch.compile, since torch.compile cannot optimize
the anything (including tensors) within the object, the object must be an
input to the graph.
You can register a custom class as being a reference-based opaque object class
through `register_opaque_type(MyClass)`.

VALUE TYPES:

Value-typed opaque objects represent constant values.
In torch.compile, the graph specializes on the object like how other constants
are. Therefore there are a couple of methods on the class that must be
implemented before registering it as a value-typed opaque object class:
  - `__eq__`: torch.compile will create guards based on the equality of this
  object, meaning that a recompilation will happen if `__eq__` returns False.
  - `__hash__`: This must be implemented for Fake Tensor caching
  - `__repr__`: This must be implemented as it will be used in the FX graph's
    codegen to reconstruct the object. The string representation must be able to
    construct the object again through its `__init__` method.
You can register a custom class as being a reference-based opaque object class
through `register_opaque_type(MyClass, value_type=True)`.

The graph containing a value-type looks like:
```
def forward(self, L_x_ : torch.Tensor):
    l_x_ = L_x_
    process_with_config = torch.ops._TestOpaqueObject.process_with_config(l_x_, ValueConfig(mode='square'));  l_x_ = None
    return (process_with_config,)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela chenyang78

[ghstack-poisoned]

angelayi committed:
Update on "[opaque obj] Add value-type opaque obj"


Opaque objects are the way we allow custom operators to accept a user-defined "black box" object as an input. There are two kinds of opaque types: VALUE type and REFERENCE type. The distinction determines how torch.compile handles the object.

REFERENCE TYPES (default):

Reference-typed opaque objects represent mutable stateful objects and are
treated as black boxes. In torch.compile, since torch.compile cannot optimize
the anything (including tensors) within the object, the object must be an
input to the graph.
You can register a custom class as being a reference-based opaque object class
through `register_opaque_type(MyClass)`.

VALUE TYPES:

Value-typed opaque objects represent constant values.
In torch.compile, the graph specializes on the object like how other constants
are. Therefore there are a couple of methods on the class that must be
implemented before registering it as a value-typed opaque object class:
  - `__eq__`: torch.compile will create guards based on the equality of this
  object, meaning that a recompilation will happen if `__eq__` returns False.
  - `__hash__`: This must be implemented for Fake Tensor caching
  - `__repr__`: This must be implemented as it will be used in the FX graph's
    codegen to reconstruct the object. The string representation must be able to
    construct the object again through its `__init__` method.
You can register a custom class as being a reference-based opaque object class
through `register_opaque_type(MyClass, value_type=True)`.

The graph containing a value-type looks like:
```
def forward(self, L_x_ : torch.Tensor):
    l_x_ = L_x_
    process_with_config = torch.ops._TestOpaqueObject.process_with_config(l_x_, ValueConfig(mode='square'));  l_x_ = None
    return (process_with_config,)
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela chenyang78

[ghstack-poisoned]
